### PR TITLE
[rolldpos] refactor rolldpos.ChainManager

### PR DESCRIPTION
**Pull request by huof6829**
# Description
Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes https://github.com/iotexproject/iotex-core/issues/3462

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Liuhaai** [comment]
Should the old interface be removed?

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -86,6 +86,10 @@ type (
 		Genesis() genesis.Genesis
 		// Context returns current context
 		Context(context.Context) (context.Context, error)
+		// BlockProposeTime return propose time by height
+		BlockProposeTime(uint64) (time.Time, error)
+		// BlockCommitTime return commit time by height
+		BlockCommitTime(uint64) (time.Time, error)
```
Comment (dustinxie) [inline blockchain/blockchain.go:92]: no need to change this interface?

File: consensus/scheme/rolldpos/rolldposctx.go [hunk 2]
```diff
@@ -510,14 +510,14 @@ func (ctx *rollDPoSCtx) Commit(msg interface{}) (bool, error) {
 
 	_consensusDurationMtc.WithLabelValues().Set(float64(time.Since(ctx.round.roundStartTime)))
 	if pendingBlock.Height() > 1 {
-		prevBlkHeader, err := ctx.chain.BlockHeaderByHeight(pendingBlock.Height() - 1)
+		prevBlkProposeTime, err := ctx.chain.BlockProposeTime(pendingBlock.Height() - 1)
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldposctx.go:513]: add internal func here
```
func (ctx *rollDPoSCtx) BlockProposeTime()
func (ctx *rollDPoSCtx) BlockCommitTime()
```

File: consensus/scheme/rolldpos/rolldpos.go [hunk 3]
```diff
@@ -36,26 +36,28 @@ var (
...
+		// CommitBlock validates and appends a block to the chain
+		CommitBlock(blk *block.Block) error
+		// ValidateBlock validates a new block before adding it to the blockchain
+		ValidateBlock(blk *block.Block) error
+		// TipHeight returns tip block's height
+		TipHeight() uint64
+		// ChainAddress returns chain address on parent chain, the root chain return empty.
+		ChainAddress() string
+	}
+)
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldpos.go:60]: add an internal struct to implement ChainManager interface
```
struct xxx {
   bc blockchain.Blockchain
}
func (c *xxx) BlockProposeTime()
func (c *xxx) BlockCommitTime()
```

File: consensus/scheme/rolldpos/rolldpos.go [hunk 4]
```diff
@@ -36,26 +36,28 @@ var (
...
-}
+type (
+	// ChainManager defines the blockchain interface
+	ChainManager interface {
+		// BlockProposeTime return propose time by height
+		BlockProposeTime(uint64) (time.Time, error)
+		// BlockCommitTime return commit time by height
+		BlockCommitTime(uint64) (time.Time, error)
+		// Genesis returns the genesis
+		Genesis() genesis.Genesis
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldpos.go:47]: according to issue, can delete `Genesis()` if it's not used

Comment (huof6829) [reply consensus/scheme/rolldpos/rolldpos.go]: Yes,

File: consensus/scheme/rolldpos/rolldpos.go [hunk 5]
```diff
@@ -39,12 +38,6 @@ var (
 type (
 	// ChainManager defines the blockchain interface
 	ChainManager interface {
-		// BlockProposeTime return propose time by height
-		BlockProposeTime(uint64) (time.Time, error)
-		// BlockCommitTime return commit time by height
-		BlockCommitTime(uint64) (time.Time, error)
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldpos.go]: still keep these 2

File: consensus/scheme/rolldpos/roundcalculator.go [hunk 6]
```diff
@@ -266,3 +268,36 @@ func (c *roundCalculator) calculateProposer(
 	proposer = delegates[idx%numDelegates]
 	return
 }
+
+type blockTime struct {
+	bc blockchain.Blockchain
+}
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/roundcalculator.go:274]: make this struct implement `ChainManager` interface

File: consensus/scheme/rolldpos/rolldpos.go (hunk 5 above)
Comment (dustinxie) [reply consensus/scheme/rolldpos/rolldpos.go]: and add `GenesisTime()`

File: consensus/scheme/rolldpos/roundcalculator.go [hunk 7]
```diff
@@ -23,6 +24,7 @@ type roundCalculator struct {
 	rp                   *rolldpos.Protocol
 	delegatesByEpochFunc DelegatesByEpochFunc
 	beringHeight         uint64
+	blockTime            *blockTime
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/roundcalculator.go:27]: no need to add this

File: consensus/scheme/rolldpos/rolldposctx.go [hunk 8]
```diff
@@ -138,6 +138,9 @@ func newRollDPoSCtx(
 		rp:                   rp,
 		timeBasedRotation:    timeBasedRotation,
 		beringHeight:         beringHeight,
+		blockTime: &blockTime{
+			bc: chain.(blockchain.Blockchain),
+		},
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldposctx.go:143]: no need to add this, modify `chain` to
```
chain: &blockTime{
			bc: chain.(blockchain.Blockchain),
		},
```

File: consensus/scheme/rolldpos/roundcalculator.go (hunk 6 above)
Comment (huof6829) [reply consensus/scheme/rolldpos/roundcalculator.go]: fixed

File: consensus/scheme/rolldpos/rolldpos.go [hunk 9]
```diff
@@ -36,25 +36,84 @@ var (
...
+		return time.Now(), errors.Wrapf(
+			err, "error when getting the block at height: %d",
+			height,
+		)
+	}
+	return footer.CommitTime(), nil
+}
+
+// GenesisTime return Genesis time by default
+func (cm *chainManager) GenesisTime() time.Time {
```
Comment (CoderZhi) [inline consensus/scheme/rolldpos/rolldpos.go:90]: if both `BlockCommitTime(0)` and `BlockProposeTime(0)` return `bc.Genesis().Timestamp`, we can delete this function from the interface.

File: consensus/scheme/rolldpos/rolldpos.go [hunk 10]
```diff
@@ -267,8 +326,8 @@ func (b *Builder) SetPriKey(priKey crypto.PrivateKey) *Builder {
 }
 
 // SetChainManager sets the blockchain APIs
-func (b *Builder) SetChainManager(chain ChainManager) *Builder {
-	b.chain = chain
+func (b *Builder) SetChainManager(bc blockchain.Blockchain) *Builder {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply consensus/scheme/rolldpos/rolldpos.go:329]: right, still use `SetChainManager(chain ChainManager)`

File: consensus/scheme/rolldpos/rolldpos.go [hunk 11]
```diff
@@ -267,8 +326,8 @@ func (b *Builder) SetPriKey(priKey crypto.PrivateKey) *Builder {
...
+	b.chain = &chainManager{bc: bc}
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldpos.go:330]: define a `newChainManager(bc blockchain.Blockchain) ChainManager`

File: consensus/scheme/rolldpos/rolldposctx_test.go [hunk 12]
```diff
@@ -44,7 +44,7 @@ func TestRollDPoSCtx(t *testing.T) {
 	})
 
 	t.Run("case 2:panic because of rp is nil", func(t *testing.T) {
-		_, err := newRollDPoSCtx(consensusfsm.NewConsensusConfig(cfg), dbConfig, true, time.Second, true, b, block.NewDeserializer(0), nil, nil, dummyCandidatesByHeightFunc, "", nil, nil, 0)
+		_, err := newRollDPoSCtx(consensusfsm.NewConsensusConfig(cfg), dbConfig, true, time.Second, true, &chainManager{bc: b}, block.NewDeserializer(0), nil, nil, dummyCandidatesByHeightFunc, "", nil, nil, 0)
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldposctx_test.go:47]: `newChainManager(b)`

File: consensus/scheme/rolldpos/rolldpos.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (huof6829) [reply consensus/scheme/rolldpos/rolldpos.go]: fixed

File: consensus/scheme/rolldpos/rolldpos.go (hunk 9 above)
Comment (huof6829) [reply consensus/scheme/rolldpos/rolldpos.go]: yes

File: consensus/scheme/rolldpos/roundcalculator.go [hunk 13]
```diff
@@ -265,3 +271,38 @@ func (c *roundCalculator) calculateProposer(
 	proposer = delegates[idx%numDelegates]
 	return
 }
+
+type blockTime struct {
+	bc blockchain.Blockchain
+}
+
+func newBlockTime(bc blockchain.Blockchain) *blockTime {
```
Comment (Liuhaai) [inline consensus/scheme/rolldpos/roundcalculator.go:279]: The arg here has to be ChainManager

File: consensus/scheme/rolldpos/rolldpos.go (hunk 11 above)
Comment (huof6829) [reply consensus/scheme/rolldpos/rolldpos.go]: right!

File: consensus/scheme/rolldpos/roundcalculator.go (hunk 13 above)
Comment (huof6829) [reply consensus/scheme/rolldpos/roundcalculator.go]: ok

File: consensus/scheme/rolldpos/rolldposctx.go [hunk 14]
```diff
@@ -511,14 +511,14 @@ func (ctx *rollDPoSCtx) Commit(msg interface{}) (bool, error) {
 
 	_consensusDurationMtc.WithLabelValues().Set(float64(time.Since(ctx.round.roundStartTime)))
 	if pendingBlock.Height() > 1 {
-		prevBlkHeader, err := ctx.chain.BlockHeaderByHeight(pendingBlock.Height() - 1)
+		prevBlkProposeTime, err := ctx.roundCalc.chain.BlockProposeTime(pendingBlock.Height() - 1)
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldposctx.go:514]: it should be `ctx.chain`, which is set by `SetChainManager()`

File: consensus/scheme/rolldpos/roundcalculator.go [hunk 15]
```diff
@@ -131,20 +130,25 @@ func (c *roundCalculator) roundInfo(
...
 				return
 			}
-			lastBlockTime = lastBlockTime.Add(lastBlock.Timestamp().Sub(lastBlockTime) / blockInterval * blockInterval)
+			lastBlockTime = lastBlockTime.Add(lastBlkProposeTime.Sub(lastBlockTime) / blockInterval * blockInterval)
 		} else {
-			var lastBlock *block.Footer
-			if lastBlock, err = c.chain.BlockFooterByHeight(height - 1); err != nil {
+			var lastBlkCommitTime time.Time
+			lastBlkCommitTime, err = c.chain.BlockCommitTime(height - 1)
+			if err != nil {
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/roundcalculator.go:148]: same as above, combine into 1 line

File: consensus/scheme/rolldpos/rolldposctx.go (hunk 14 above)
Comment (huof6829) [reply consensus/scheme/rolldpos/rolldposctx.go]: fixed

File: consensus/scheme/rolldpos/roundcalculator.go [hunk 16]
```diff
@@ -131,20 +130,25 @@ func (c *roundCalculator) roundInfo(
...
+	if lastBlockTime, err = c.chain.BlockProposeTime(0); err != nil {
+		return
+	}
 	if height > 1 {
 		if height >= c.beringHeight {
-			var lastBlock *block.Header
-			if lastBlock, err = c.chain.BlockHeaderByHeight(height - 1); err != nil {
+			var lastBlkProposeTime time.Time
+			lastBlkProposeTime, err = c.chain.BlockProposeTime(height - 1)
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (huof6829) [reply consensus/scheme/rolldpos/roundcalculator.go]: fixed

File: consensus/scheme/rolldpos/roundcalculator.go (hunk 15 above)
Comment (huof6829) [reply consensus/scheme/rolldpos/roundcalculator.go]: fixed

**Review by dustinxie — APPROVED** [review]

File: consensus/scheme/rolldpos/rolldposctx.go [hunk 17]
```diff
@@ -511,14 +511,14 @@ func (ctx *rollDPoSCtx) Commit(msg interface{}) (bool, error) {
 
 	_consensusDurationMtc.WithLabelValues().Set(float64(time.Since(ctx.round.roundStartTime)))
 	if pendingBlock.Height() > 1 {
-		prevBlkHeader, err := ctx.chain.BlockHeaderByHeight(pendingBlock.Height() - 1)
+		prevBlkProposeTime, err := ctx.chain.BlockProposeTime(pendingBlock.Height() - 1)
 		if err != nil {
 			log.L().Error("Error when getting the previous block header.",
```
Comment (Liuhaai) [inline consensus/scheme/rolldpos/rolldposctx.go:516]: logging info needs to be adjusted as well

File: consensus/scheme/rolldpos/rolldpos.go [hunk 18]
```diff
@@ -36,25 +36,87 @@ var (
...
+}
+
+// BlockProposeTime return propose time by height
+func (cm *chainManager) BlockProposeTime(height uint64) (time.Time, error) {
+	if height == 0 {
+		return time.Unix(cm.bc.Genesis().Timestamp, 0), nil
+	}
+	header, err := cm.bc.BlockHeaderByHeight(height)
+	if err != nil {
+		return time.Now(), errors.Wrapf(
```
Comment (Liuhaai) [inline consensus/scheme/rolldpos/rolldpos.go:77]: time.Now() is not default value for `Time`.

File: consensus/scheme/rolldpos/rolldposctx.go (hunk 17 above)
Comment (huof6829) [reply consensus/scheme/rolldpos/rolldposctx.go]: ok

File: consensus/scheme/rolldpos/rolldpos.go (hunk 18 above)
Comment (huof6829) [reply consensus/scheme/rolldpos/rolldpos.go]: fixed

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**